File: sks/sks/custom/py/delivery_trip.py
import frappe
import json
from datetime import date
from frappe.desk.form import assign_to

@frappe.whitelist()
def update_invoice(invoice,fields):
    fields = json.loads(fields)
    frappe.db.set_value("Sales Invoice",invoice, 'delivery_status', fields.get('delivery_status'))
    frappe.db.set_value("Sales Invoice",invoice, 'reason', fields.get('reason'))
    frappe.db.set_value("Sales Invoice",invoice, 'time_of_delivery', fields.get('time_of_delivery'))
    frappe.db.commit()
    return ""

@frappe.whitelist()
def payment_entry(mode,amount,pending_invoice,company):
    customer = frappe.get_value("Sales Invoice",pending_invoice,'customer')
    outstanding = frappe.get_value("Sales Invoice",pending_invoice, 'outstanding_amount')
    ref_date = date.today()
    ref_no = "12315"
    mode_of_payment = frappe.get_doc("Mode of Payment",mode).accounts
    for i in mode_of_payment:
        if(i.company==company):
            acc_paid_to=i.default_account
            break
    try:
        if(acc_paid_to):pass
    except:
        frappe.throw(("Please set Company and Default account for ({0}) mode of payment").format("<a href = /app/mode-of-payment/"+"%20".join(mode.split())+">"+mode+"</a href>"))
    bank_account_type = frappe.db.get_value("Account", acc_paid_to, "account_type")
    if bank_account_type == "Bank":
        if(ref_no == None or ref_date == None):
            frappe.throw("Reference No and Reference Date is mandatory for Bank transaction")
    acc_currency = frappe.db.get_value('Account',acc_paid_to,'account_currency')
    doc = frappe.new_doc('Payment Entry')
    references=[]
    amount1 = float(amount)
    if(amount1 > outstanding):
        return "Unable to create payment entry for {0}".format(pending_invoice), 'red',amount1,outstanding
        
    references.append({
        'reference_doctype':'Sales Invoice',
        'reference_name': pending_invoice,
        'total_amount':amount1,
        'exchange_rate': 1,
        'allocated_amount': amount1
    })
    doc.update({
        'company':company,
        'payment_type':"Receive",
        'docstatus': 1,
        'mode_of_payment':mode,
        'party_type': 'Customer',
        'party': customer,
        'paid_amount':float(amount),
        'source_exchange_rate':1,
        'references':references,
        'received_amount':float(amount),
        'target_exchange_rate':1,
        'paid_to': acc_paid_to,
        'paid_to_account_currency': acc_currency,
        # 'pos_opening_shift_id': opening
    })
    if(bank_account_type == 'Bank'):
        doc.update({
            'reference_no':ref_no,
            'reference_date':ref_date
        })
    doc.insert()
    doc.submit()
    frappe.db.commit()
    return "Payment Entry {0} created for {1}".format(doc.name,pending_invoice),'green'

@frappe.whitelist()
def get_customer_territory():
    return frappe.get_all("Territory",filters={'is_group':0},pluck="name") 

@frappe.whitelist()
def get_condition_from_dialog(data):
    data = json.loads(data)
    filter={'docstatus':1}
    data_key = list(data.keys())
    if('customer' in data_key):
        filter['customer'] = data['customer']
    if('day' in data_key):
        filter['delivery_day'] = (data['day'])
    if('territory' in data_key):
        filter['territory'] = ['in', data['territory']]
    sales_order = get_sales_order(filter)

    filter1 = {}
    sales_invoice =[]
    if(len(sales_order)):
        filter1['sales_order'] = ['in',sales_order]
        if('invoice' in data_key):
            filter1['parent'] = data['invoice']
        # filter1 = {'delivery_status': ['not in', ['Delivered', 'Returned']]}
        sales_invoice = get_sales_invoice(filter1, 'Sales Invoice Item','parent')
    if(len(sales_invoice)):
        if('outstanding' in data_key):
            filter1['outstanding_amount'] = ['>=', data['outstanding']]
            filter1['name'] = ['in', sales_invoice]
            filter1['docstatus'] = 1
            filter1['delivery_status']= ['not in', ['Delivered', 'Returned','Out to Delivery']]
            sales_invoice = get_sales_invoice(filter1,"Sales Invoice",'name')
            return get_sales_invoice_details(sales_invoice)
    else:
        frappe.throw("Not any Sales Invoice matched for this condition")

def get_sales_order(filter):
    return frappe.get_all("Sales Order",filters=filter,pluck="name")

def get_sales_invoice(filter,doctype,name):
    return frappe.get_all(doctype, filters= filter, pluck = name)

def get_sales_invoice_details(sales_invoice):
    return frappe.get_all("Sales Invoice",fields=['name','customer_address','customer','delivery_status','reason','contact_person','time_of_delivery','rounded_total','outstanding_amount'],filters ={'name':['in',sales_invoice],'delivery_status':['not in', ['Delivered', 'Returned']]})
def assign_to_driver(del_trip,a):

    # Updation of Out to Delivery Status
    for i in del_trip.delivery_stops:
        sales_inv_doc = frappe.get_doc('Sales Invoice',i.sales_invoice)
        sales_inv_doc.delivery_status = 'Out to Delivery'
        sales_inv_doc.save()

    allow=0
    row=0
    for i in del_trip.delivery_stops:
        if(i.delivery_status != "Delivered"):
            allow=1
    if(allow==1):
        doc = frappe.new_doc('TS Driver Delivery Trip')
        del_trip_doc = frappe.get_doc('Delivery Trip', del_trip.name)
        doc.update({
            'driver_id':del_trip_doc.driver,
            'driver': del_trip_doc.driver_name,
            'vehicle': del_trip_doc.vehicle,
            'delivery_trip': del_trip_doc.name,
            'company': del_trip_doc.company,
            'status': "Open"
        })
        doc.insert(ignore_permissions=True)
        if(del_trip.driver):
            assign_to.add(
                dict(
                    assign_to=[del_trip.user_id],
                    doctype="TS Driver Delivery Trip",
                    name=doc.name,
                    notify=True,
                )
            )
        return doc
    else:
        frappe.throw("All the Sales Invoice are Delivered")

def update_sales_invoice(doc,event):
    # Updation of Hold Status
    for i in doc.delivery_stops:
        sales_inv_doc = frappe.get_doc('Sales Invoice',i.sales_invoice)
        sales_inv_doc.delivery_status = 'Hold'
        sales_inv_doc.save()

File: sks/sks/custom/js/delivery_trip.js
let filter;
let company;
frappe.ui.form.on('Delivery Trip',{
    refresh: function(frm,cdt,cdn){
        company = frm.doc.company;
        if(frm.doc.docstatus == 0){
        frm.add_custom_button( __("Sales Invoice"),function(){
            frappe.call({
                'method':"sks.sks.custom.py.delivery_trip.get_customer_territory",
                'args': {},
                callback(r){

                    var d = new frappe.ui.Dialog({
                        title: "Choose Invoices For Delivery 🛻🛻🛻🛻..",
                        fields: [
                            {label:'Customer',fieldname:'customer',fieldtype:'Link',options: 'Customer'},
                            {fieldtype:'Column Break'},
                            {label:'Invoice',fieldname:'invoice',fieldtype:'Link',options: 'Sales Invoice'},
                            {fieldtype:'Column Break'},
                            {label:'Outstanding greater than',fieldname:'outstanding',fieldtype:'Float'},
                            {label:'Detlivery date between',fieldname:'sec',fieldtype:'Section Break'},
                            {label:'From Date',fieldname:'from_date',fieldtype:'Date'},
                            {fieldtype:'Column Break'},
                            {label:'To Date',fieldname:'to_date',fieldtype:'Date'},
                            {label:'Select territory',fieldname:'sec1',fieldtype:'Section Break'},
                            {label:'Territory',fieldname:'territory',fieldtype: 'MultiSelectPills',
                            get_data: function() {
                                return r.message
                            }
                        }
                        ],
                        primary_action_label: "Submit",
                        primary_action: function(data){
                            filter = data
                            frappe.call({
                                method: "sks.sks.custom.py.delivery_trip.get_condition_from_dialog",
                                args: {
                                    data: data
                                },
                                callback(r){
<<<<<<< HEAD
                                        if(r.message.length == 0){
                                            frappe.throw({
                                                title: "No Invoice",
                                                message: "All Invoice are Deliver"
                                             })
                                        }
=======
>>>>>>> e7ed9646e8968c26748de06d44d007471731dd3e
                                    let p = locals[cdt][cdn]
                                    frm.set_value("delivery_stops",[])
                                        for(let i = 0; i<r.message.length;i++){
                                        cur_frm.add_child("delivery_stops")
                                        frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"customer",r.message[i]["customer"])
                                        frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"address",r.message[i]["customer_address"])
                                        frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"sales_invoice",r.message[i]["name"])
                                        frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"delivery_status",r.message[i]["delivery_status"])
                                        frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"reason",r.message[i]["reason"])
                                        frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"contact",r.message[i]["contact_person"])
                                        frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"time_of_delivery",r.message[i]["time_of_delivery"])
                                        frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"amount",r.message[i]["outstanding_amount"])
                                         }
                                        cur_frm.refresh_field('delivery_stops')


                                }
                            })
                            d.hide();
                            
                        }
                    })
                    d.show()
                }
            })
            
        }, __('Get customers from'))
    }
    },
    update_invoice: function(frm,cdt,cdn){
        for(let i=0;i<2;i++){
        if(frm.doc.sales_invoice){
            frappe.call({
                method: "sks.sks.custom.py.delivery_trip.update_invoice",
                args:{
                    invoice: frm.doc.sales_invoice,
                    fields:{
                        delivery_status: frm.doc.delivery_status,
                        reason:frm.doc.reason || "",
                        time_of_delivery: frm.doc.time_of_delivery 
                    },
                    callback(res){
                        cur_frm.refresh_field('delivery_stops')
                        frm.refresh()
                        let p = locals[cdt][cdn]
                                
                                frappe.call({
                                    method : "sks.sks.custom.py.delivery_trip.get_condition_from_dialog",
                                    args : {data: filter},
                                    callback(r){
                                        frm.set_value("delivery_stops",[])
                                            for(let i = 0; i<r.message.length;i++){
                                            cur_frm.add_child("delivery_stops")
                                            frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"customer",r.message[i]["customer"])
                                            frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"address",r.message[i]["customer_address"])
                                            frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"sales_invoice",r.message[i]["name"])
                                            frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"delivery_status",r.message[i]["delivery_status"])
                                            frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"reason",r.message[i]["reason"])
                                            frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"contact",r.message[i]["contact_person"])
                                            frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"time_of_delivery",r.message[i]["time_of_delivery"])
                                            frappe.model.set_value(p.delivery_stops[i].doctype,p.delivery_stops[i].name,"amount",r.message[i]["outstanding_amount"])
                                            cur_frm.refresh_field('delivery_stops')
                                            
                                        }
                                        cur_frm.refresh_field('delivery_stops')
                                        frm.refresh()
                                    }
                                })

                    }
                } 
            })
        }
    }
    //payment entry

    if(frm.doc.mode_of_payment && frm.doc.amount && frm.doc.sales_invoice && frm.doc.delivery_status == "Delivered"){
        frappe.call({
            method:"sks.sks.custom.py.delivery_trip.payment_entry",
            args:{
                mode: frm.doc.mode_of_payment,
                amount: frm.doc.amount,
                pending_invoice: frm.doc.sales_invoice,
                company: frm.doc.company,
            },
            callback(res){
                frappe.show_alert({
                    message: res.message[0],
                    indicator: res.message[1]
                });
                if(res.message[1] == 'red'){
                    frappe.throw({
                       title: "Amount Exceed",
                       message: "Paid Amount is greater than the Outstanding Amount. ("+res.message[2]+" > "+res.message[3]+")"
                    })
                }
            }
        })
    }
},
    sales_invoice: function(frm){
        let invoices=[];
        for(let i of frm.doc.delivery_stops){
            invoices.push(i.sales_invoice)
        }
        frm.set_query("sales_invoice", function(){
            return{
                filters: {
                    'name': ['in',invoices]
                }
            }
        })
    }

})

frappe.ui.form.on("Delivery Stop",{
    'detail': function(frm,cdt,cdn){
        let p = locals[cdt][cdn]
        if(p.sales_invoice){
        function update_fields(p,r) {
            let f = locals[p.parenttype][p.parent]
                        frm.set_value("delivery_stops",[])
                            for(let i = 0; i<r.message.length;i++){
                            cur_frm.add_child("delivery_stops")
                            frappe.model.set_value(f.delivery_stops[i].doctype,f.delivery_stops[i].name,"customer",r.message[i]["customer"])
                            frappe.model.set_value(f.delivery_stops[i].doctype,f.delivery_stops[i].name,"address",r.message[i]["customer_address"])
                            frappe.model.set_value(f.delivery_stops[i].doctype,f.delivery_stops[i].name,"sales_invoice",r.message[i]["name"])
                            frappe.model.set_value(f.delivery_stops[i].doctype,f.delivery_stops[i].name,"delivery_status",r.message[i]["delivery_status"])
                            frappe.model.set_value(f.delivery_stops[i].doctype,f.delivery_stops[i].name,"reason",r.message[i]["reason"])
                            frappe.model.set_value(f.delivery_stops[i].doctype,f.delivery_stops[i].name,"contact",r.message[i]["contact_person"])
                            frappe.model.set_value(f.delivery_stops[i].doctype,f.delivery_stops[i].name,"time_of_delivery",r.message[i]["time_of_delivery"])
                            frappe.model.set_value(f.delivery_stops[i].doctype,f.delivery_stops[i].name,"amount",r.message[i]["outstanding_amount"])
                                }
                            cur_frm.refresh_field('delivery_stops')
        }

        let reason=p.reason
        if(reason == ""){reason = " "}
        let amount=p.amount
        if(amount == 0){amount = "0"}
        var today = new Date();
        var time = today.getHours() + ":" + today.getMinutes() + ":" + today.getSeconds();
        var d = new frappe.ui.Dialog({
            title: "Invoice: " + p.sales_invoice,
            fields: [
            {fieldname: 'sales_invoice',label: 'Invoice',fieldtype: 'Link', default:p.sales_invoice, read_only: 1,options:'Sales Invoice'},
            {fieldtype:'Column Break'},
            {fieldname: 'customer',label: 'Customer',fieldtype: 'Link', default:p.customer, read_only: 1,options:'Customer'},
            {fieldtype:'Column Break'},
            {fieldname: 'dis_amount',label: 'Amount',fieldtype: 'Currency', default:amount, read_only: 1},
            {fieldtype:'Section Break'},
            {fieldname: 'address',label: 'Address Name',fieldtype: 'Link', default:p.address, read_only: 1,options:'Address'},
            {fieldtype:'Column Break'},
            {fieldname: 'dis_reason',label: 'Reason',fieldtype: 'Data', default:reason, read_only: 1},
            {fieldtype:'Section Break',label: "Update Status"},
            {fieldname: 'delivery_status',label: 'Delivery Status',fieldtype: 'Select', options: ['Attempt','Delivered','Not Delivered','Ready To Dispatch', 'Reattempt,Returned'], default:p.delivery_status},
            {fieldtype:'Column Break'},
            {fieldname: 'reason',label: 'Reason',fieldtype: 'Data', default:reason},
            {fieldtype:'Section Break'},
            {fieldname: 'mode_of_payment',label: 'Mode of Payment',fieldtype: 'Link',options: 'Mode of Payment',default: 'Cash'},
            {fieldtype:'Column Break'},
            {fieldname: 'amount',label: 'Paid Amount',fieldtype: 'Currency'},
            {fieldtype:'Column Break'},
            {fieldname: 'time_of_delivery',label: 'Time of Delivery',fieldtype: 'Time', default:time},
            {fieldtype:'Section Break'},

        ],
        primary_action_label: "Update Invoice",
        primary_action: function(data){
             //payment entry

   if(data.mode_of_payment && data.amount && data.sales_invoice && data.delivery_status == "Delivered"){
    frappe.call({
        method:"sks.sks.custom.py.delivery_trip.payment_entry",
        args:{
            mode: data.mode_of_payment,
            amount: data.amount,
            pending_invoice: data.sales_invoice,
            company: company,
        },
        callback(res){
<<<<<<< HEAD
            // frappe.call({
            //     method: "sks.sks.custom.py.delivery_trip.get_condition_from_dialog",
            //     args: {
            //         data: filter
            //     },
            //     callback(r){
            //         update_fields(p,r)


            //     }
            // })
=======
            frappe.call({
                method: "sks.sks.custom.py.delivery_trip.get_condition_from_dialog",
                args: {
                    data: filter
                },
                callback(r){
                    update_fields(p,r)


                }
            })
>>>>>>> e7ed9646e8968c26748de06d44d007471731dd3e
            frappe.show_alert({
                message: res.message[0],
                indicator: res.message[1]
            });
            if(res.message[1] == 'red'){
                frappe.throw({
                   title: "Amount Exceed",
                   message: "Paid Amount is greater than the Outstanding Amount. ("+res.message[2]+" > "+res.message[3]+")"
                })
            }
        }
    })
}
            frappe.call({
                method: "sks.sks.custom.py.delivery_trip.update_invoice",
                args:{
                    invoice: p.sales_invoice,
                    fields: data
                },
                callback(res){
                frappe.call({
                    method: "sks.sks.custom.py.delivery_trip.get_condition_from_dialog",
                    args: {
                        data: filter
                    },
                    callback(r){
                        update_fields(p,r)


                    }
                })
                }
            })
  
            d.hide();
    }
        })
        d.show()
    }}
})